Mathlib/Combinatorics/Enumerative/Stirling.lean

```diff
@@ -0,0 +1,210 @@
+  simp [choose_two_right]
+  induction' n with n hn
+  · simp [stirlingFirst]
+  · have h₀ : (n + 1) = (2 * (n + 1)) / 2 := by
+      rw [mul_comm, Nat.mul_div_assoc, Nat.div_self, mul_one]
+      · omega
+      · exact Nat.dvd_refl _
+    rw [stirlingFirst_succ_succ, hn, stirlingFirst_self, mul_one]
+    nth_rw 1 [h₀]
+    rw [← Nat.add_div_of_dvd_left]
+    · ring_nf
+    · suffices h₁ : Even (n * (n + 1)) from by
+        rw [even_iff_two_dvd, mul_comm] at h₁
+        exact h₁
+      exact Nat.even_mul_succ_self n
+
+theorem stirlingFirst_one_right (n : ℕ) : stirlingFirst (n + 1) 1 = n.factorial := by
+  induction' n with n hn
+  · simp [stirlingFirst]
+  · rw [stirlingFirst_succ_succ, zero_add, hn, stirlingFirst_succ_zero]
+    simp [Nat.sub_self, Nat.factorial_succ]
+
+
+/--
+`Nat.stirlingSecond n k` is the Stirling number of the second kind,
+counting the number of ways to partition a set of `n` elements into `k` nonempty subsets.
+-/
+def stirlingSecond : ℕ → ℕ → ℕ
+  | 0, 0 => 1
+  | 0, _ + 1 => 0
+  | _ + 1, 0 => 0
+  | n + 1, k + 1 =>
+    (k + 1) * stirlingSecond n (k + 1) + stirlingSecond n k
+
+@[simp]
+theorem stirlingSecond_zero : stirlingSecond 0 0 = 1 :=
+  rfl
+
+@[simp]
+theorem stirlingSecond_zero_succ (k : ℕ) : stirlingSecond 0 (succ k) = 0 :=
+  rfl
+
+@[simp]
+theorem stirlingSecond_zero_right' (n : ℕ) : stirlingSecond (succ n) 0 = 0 :=
+  rfl
+
+theorem stirlingSecond_succ_left (n k : ℕ) (hk : k ≠ 0) :
+    stirlingSecond (n + 1) k = k * stirlingSecond n k + stirlingSecond n (k - 1) := by
+  have hk : 0 < k := Nat.pos_of_ne_zero hk
+  obtain ⟨l, rfl⟩ : ∃ l, k = l + 1 := Nat.exists_eq_add_of_le' hk
+  rfl
+
+theorem stirlingSecond_succ_right (n k : ℕ) (hn : n ≠ 0) :
+    stirlingSecond n (k + 1) =
+     (k + 1) * stirlingSecond (n - 1) (k + 1) + stirlingSecond (n - 1) k := by
+  have hn : 0 < n := Nat.pos_of_ne_zero hn
+  obtain ⟨l, rfl⟩ : ∃ l, n = l + 1 := Nat.exists_eq_add_of_le' hn
+  rfl
+
+theorem stirlingSecond_succ_succ (n k : ℕ) :
+    stirlingSecond (n + 1) (k + 1) =
+     (k + 1) * stirlingSecond n (k + 1) + stirlingSecond n k := by
+  rw [stirlingSecond]
+
+theorem stirlingSecond_eq_zero_of_lt : ∀ {n k : ℕ}, n < k → stirlingSecond n k = 0
+  | _, 0, hk => absurd hk (Nat.not_lt_zero _)
+  | 0, _ + 1, _ => by rw [stirlingSecond]
+  | n + 1, k + 1, hk => by
+    have hnk : n < k := Nat.lt_of_succ_lt_succ hk
+    have hnk1 : n < k + 1 := Nat.lt_of_succ_lt hk
+    rw [stirlingSecond_succ_succ, stirlingSecond_eq_zero_of_lt hnk,
+     stirlingSecond_eq_zero_of_lt hnk1]
+    simp
+
+theorem stirlingSecond_self (n : ℕ) : stirlingSecond n n = 1 := by
+  induction n <;> simp [*, stirlingSecond, stirlingSecond_eq_zero_of_lt (lt_succ_self _)]
+
+theorem stirlingSecond_one_right (n : ℕ) : stirlingSecond (n + 1) 1 = 1 := by
+  simp [stirlingSecond]
+  induction' n with n ih
+  · simp [stirlingSecond]
+  · rw [stirlingSecond_zero_right']
+    simp
```

Comment:
This is a non-terminal simp, which should be avoided in mathlib, see https://leanprover-community.github.io/extras/simp.html#non-terminal-simps
There are a couple like this in the file, please make sure to adjust.